File: AlgorithmsJS/QuickSort.js
class SortableArray {
  constructor(array) {
    this.array = array;
  }

  swap(pointer1, pointer2) {
    const tmpValue = this.array[pointer1];
    this.array[pointer1] = this.array[pointer2];
    this.array[pointer2] = tmpValue;
  }

  setPartition(leftPointer, rightPointer) {
    this.pivotPosition = rightPointer;
    this.pivot = this.array[this.pivotPosition];

    rightPointer -= 1;

    while(true) {

      while(this.array[leftPointer] < this.pivot) {
        leftPointer += 1;
      }

      while(this.array[rightPointer] > this.pivot) {
        rightPointer -= 1;
      }

      if (leftPointer >= rightPointer) break;
      this.swap(leftPointer, rightPointer);
    }
    
    this.swap(leftPointer, this.pivotPosition);
    return leftPointer;
  }

  quickSort(leftIndex, rightIndex) {

    if(rightIndex - leftIndex <= 0) return;

    const pivotPosition = this.setPartition(leftIndex, rightIndex);
    this.quickSort(leftIndex, pivotPosition - 1);
    this.quickSort(pivotPosition + 1, rightIndex);
  }
}

module.exports = SortableArray;

File: AlgorithmsPY/quick_sort.py
class QuickSort:
  def __init__(self, array):
    self.array = array


  def swap(self, pointer1, pointer2):
    self.array[pointer1], self.array[pointer2] = self.array[pointer2], self.array[pointer1]

  def set_partition(self, left_pointer, right_pointer):
    self.pivot_position = right_pointer
    self.pivot = self.array[self.pivot_position]

    right_pointer = right_pointer - 1

    while(True):
      
      while(self.array[left_pointer] < self.pivot):
        left_pointer = left_pointer + 1

      while(self.array[right_pointer] > self.pivot):
        right_pointer = right_pointer - 1

      if left_pointer >= right_pointer:
        break
      else:
        self.swap(left_pointer, right_pointer)

    self.swap(left_pointer, self.pivot_position)
    return left_pointer

  def quick_sort(self, left_index, right_index):
    if right_index - left_index <= 0:
      return
    
    pivot_position = self.set_partition(left_index, right_index)
    self.quick_sort(left_index, pivot_position -1)
    self.quick_sort(pivot_position + 1, right_index)
